### PR TITLE
GraphQL Custom Schema

### PR DESCRIPTION
**Pull request by davimacedo**
This PR empowers the Parse GraphQL API with custom user-defined schema. The developers can now write their own types, queries, and mutations, which will merged with the ones that are automatically generated. The new types are resolved by the application's cloud code functions.

Therefore, regarding https://github.com/parse-community/parse-server/issues/5777, this PR closes the cloud functions needs and also addresses the graphql customization topic. In my view, I think that this PR, together with https://github.com/parse-community/parse-server/pull/5782 and https://github.com/parse-community/parse-server/pull/5818, when merged, closes the issue.

How it works:

1. When initializing ParseGraphQLServer, now the developer can pass a custom schema that will be merged to the auto-generated one:
```
      parseGraphQLServer = new ParseGraphQLServer(parseServer, {
        graphQLPath: '/graphql',
        graphQLCustomTypeDefs: gql`
          extend type Query {
            custom: Custom @namespace
          }
           type Custom {
            hello: String @resolve
            hello2: String @resolve(to: "hello")
            userEcho(user: _UserFields!): _UserClass! @resolve
          }
        `,
      });
```

Note:
- This PR includes a @namespace directive that can be used to the top level field of the nested queries and mutations (it basically just returns an empty object);
- This PR includes a @resolve directive that can be used to notify the Parse GraphQL Server to resolve that field using a cloud code function. The `to` argument specifies the function name. If the `to` argument is not passed, the Parse GraphQL Server will look for a function with the same name of the field;
- This PR allows creating custom types using the auto-generated ones as in `userEcho(user: _UserFields!): _UserClass! @resolve`;
- This PR allows to extend the auto-generated types, as in `extend type Query { ... }`.

2. Once the schema was set, you just need to write regular cloud code functions:
```
      Parse.Cloud.define('hello', async () => {
        return 'Hello world!';
      });

      Parse.Cloud.define('userEcho', async req => {
        return req.params.user;
      });
```

3. Now you are ready to play with your new custom api:
```
query {
  custom {
    hello
    hello2
    userEcho(user: { username: "somefolk" }) {
      username
    }
  }
}
```
should return
```
{
  "data": {
    "custom": {
      "hello": "Hello world!",
      "hello2": "Hello world!",
      "userEcho": {
        "username": "somefolk"
      }
    }
  }
}
```

### CODE REVIEW

**Comment by davimacedo** [comment]
@omairvaiyani let me know your feedback.

File: src/GraphQL/loaders/schemaDirectives.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+export const definitions = gql`
+  directive @namespace on FIELD_DEFINITION
+  directive @resolve(to: String) on FIELD_DEFINITION
+`;
+
+const load = parseGraphQLSchema => {
+  parseGraphQLSchema.graphQLSchemaDirectivesDefinitions = definitions;
+
+  class NamespaceDirectiveVistor extends SchemaDirectiveVisitor {
```
Comment (douglasmuraoka) [inline src/GraphQL/loaders/schemaDirectives.js]: Typo `NamespaceDirectiveVisitor`?

Comment (davimacedo) [reply src/GraphQL/loaders/schemaDirectives.js]: done

**Comment by ciekawy** [comment]
This is so great to saw custom schema support already there. I am wondering now... I have my own integration in place for about a year using just apollo server mounted in express along with parse. Thus I have regular resolvers. Moving stuff into nested namespace does not seem to be a problem. Just wondering if it is now possible or maybe possible to add support for regular resolvers?

**Comment by davimacedo** [comment]
@ciekawy the currently approach is using cloud functions as resolvers. When you say regular resolvers, what are you more concerned about in comparison to using cloud functions?

**Comment by ciekawy** [comment]
as for regular resolvers my main point is that as I am already using regular `ApolloServer` with `schema` and `resolvers`. In addition as I am using TypeScript, `resolvers` are of type [`IResolvers`  from `graphql-tools`](https://github.com/apollographql/graphql-tools/blob/master/docs/source/resolvers.md) and also I can easily return a reference to another resolver when it complies with the schema as well as have nested resolvers and return data depending on parent object (basically relying on `IFieldResolver` of `(parent, args, context, info) => any`).

Ideally I'd like to use Apollo compliant resolvers without (or with minimal) changes. One option I was thinking about could be to do schema stiching and delegate internally `custom: Custom @namespace` to another instance of parse server (but then the drawback most probably would be that those resolvers will be executed out of parse context unless the context could be passed somehow as a context param... ;)

**Comment by ciekawy** [comment]
so to recap possible concerns are
- preserve `IResolvers` as much as possible
- at least preserve `IFieldResolver` methods (maybe it would be possible to wrap/map them to/from the cloud functions)
- keep the flexibility of GraphQL resolvers and be able to define nested resolvers (having access to parent object) and being able to delegate to other resolvers for nested data.

Definitely I need to find time to play a bit with latest parse and at least try to understand better how the integrations works.

I see mapping custom schema to cloud function really great idea especially for those who were using cloud functions in the past and are willing to jump into the GraphQL. OTOH I think ideally there should be more natural GraphQL integration for those working already more with GraphQL then parse.

**Comment by davimacedo** [comment]
I think that we can think in some way to allow the developer to pass an executable schema (created with graphql-tools) to the Parse GraphQL Server that will be merged with the automatic generated one.

We can also pass to the cloud code functions all `(parent, args, context, info)` args for it to resolve the field.

I think both ideas would help you, right?

**Comment by ciekawy** [comment]
yep, first one would be awesome. Second would be great anyway - even for people using just the cloud functions approach as they would use the GraphQL approach from the get-go (main danger would be naming conflict yet renaming on my side is acceptable as I am now the only consumer of my schema).

I'd be very happy to contribute in this field, the only thing is thats part of my full-stack/mobile side-project when I have a stretch goal to make first MVP release :) I feel like my back-end part need some love and migrating to native parse-graphql support is the way it should be done. But now I could imagine that only by minimum effort while switching to parse 3.7.x+

**Comment by ciekawy** [comment]
looks like federation is the current way of merging schemas 
https://www.apollographql.com/docs/apollo-server/federation/migrating-from-stitching/

**Comment by davimacedo** [comment]
Nice @ciekawy ! Thanks for sharing your thoughts. We have a todo list of what we want to do in GraphQL [here](https://github.com/parse-community/parse-server/projects/5) and I've just added these two issues. You are more than welcome to contribute on this! Just pick up one of the tasks, and send as a PR. I will be happy to review.

**Pull request merged**